### PR TITLE
[ISSUE-11725] Add secondary statusCode messages on error

### PR DESCRIPTION
**Pull request by youngkih**
This fixes https://github.com/spring-projects/spring-security/issues/11725 and reflects the feedback from https://github.com/spring-projects/spring-security/pull/12818

### CODE REVIEW

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java [hunk 1]
```diff
@@ -404,6 +412,26 @@ public static Converter<ResponseToken, Saml2ResponseValidatorResult> createDefau
...
+		if (response.getStatus().getStatusCode() == null) {
+			return StatusCode.SUCCESS;
+		}
+
+		StatusCode parentStatusCode = response.getStatus().getStatusCode();
+		String parentStatusCodeValue = parentStatusCode.getValue();
+		if (includeChildStatusCodes.contains(parentStatusCodeValue)) {
+			return Optional.ofNullable(parentStatusCode.getStatusCode())
+					.map(StatusCode::getValue)
+					.map(childStatusCodeValue -> parentStatusCodeValue + childStatusCodeValue)
```
Comment (jzheaux) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java:428]: I think concatenating the messages is something more of a concern for the view. Would you please change `getStatusCode` to something like `getStatusCodes` and return a list of strings? Then the result object should add a `Saml2Error` for each one.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java [hunk 2]
```diff
@@ -404,6 +412,26 @@ public static Converter<ResponseToken, Saml2ResponseValidatorResult> createDefau
...
+			return StatusCode.SUCCESS;
+		}
+		if (response.getStatus().getStatusCode() == null) {
+			return StatusCode.SUCCESS;
+		}
+
+		StatusCode parentStatusCode = response.getStatus().getStatusCode();
+		String parentStatusCodeValue = parentStatusCode.getValue();
+		if (includeChildStatusCodes.contains(parentStatusCodeValue)) {
+			return Optional.ofNullable(parentStatusCode.getStatusCode())
```
Comment (jzheaux) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java:426]: Please don't use `Optional` for now due to https://github.com/spring-projects/spring-security/issues/7155

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java [hunk 3]
```diff
@@ -404,6 +412,26 @@ public static Converter<ResponseToken, Saml2ResponseValidatorResult> createDefau
...
+		if (response.getStatus() == null) {
+			return StatusCode.SUCCESS;
+		}
+		if (response.getStatus().getStatusCode() == null) {
+			return StatusCode.SUCCESS;
+		}
+
+		StatusCode parentStatusCode = response.getStatus().getStatusCode();
+		String parentStatusCodeValue = parentStatusCode.getValue();
+		if (includeChildStatusCodes.contains(parentStatusCodeValue)) {
```
Comment (jzheaux) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java:425]: The build is going to require that it be `this.includeChildStatusCodes`. Will you please run `./gradlew format && gw checkstyleMain checkstyleTest` on the code and make any necessary modifications?

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java [hunk 4]
```diff
@@ -51,16 +51,20 @@
 import org.opensaml.saml.saml2.core.EncryptedAssertion;
 import org.opensaml.saml.saml2.core.EncryptedAttribute;
 import org.opensaml.saml.saml2.core.EncryptedID;
+import org.opensaml.saml.saml2.core.Issuer;
```
Comment (jzheaux) [inline saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java:54]: On each file that you modify, will you please change any copyright message of the form `20xx-20yy` copyright date to `20xx-2024`?

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java [hunk 5]
```diff
@@ -729,6 +735,77 @@ public void authenticateWhenCustomResponseValidatorThenUses() {
 		verify(validator).convert(any(OpenSaml4AuthenticationProvider.ResponseToken.class));
 	}
 
+	@Test
+	public void setsOnlyParentStatusCodeOnResultDescription() {
```
Comment (jzheaux) [inline saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java:739]: Will you please name your tests like the following: `authenticateWhenXXXThenYYY` where `XXX` is the situation under test and `YYY` is the expected result?

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks, @youngkih! I've left some feedback inline.

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java (hunk 4 above)
Comment (youngkih) [reply saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProviderTests.java:54]: Fixed! Thanks!

**Review by jzheaux — APPROVED** [review]

**Comment by jzheaux** [comment]
Thanks, @youngkih! This is now merged into `main`. I also added a [polish commit](3f116226871045c9bd80f3e769f9703fdfc0e5b4) to make the code style a bit more consistent with the rest of Spring Security.

**Comment by youngkih** [comment]
Cool, thanks @jzheaux !

**Pull request merged**